src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -314,11 +322,21 @@ protected void ClearSolution()
         /// Call the base method at the end of your method.
```

Comment:
The docs for this method say "this method is called when a solution is cleared", which I guess means we call it in a response to ClearSolution -- but that's not happening now. Frankly, I'm not sure why we have two methods here, but it almost feels like this ClearSolutionData method is kinda implicitly a _NoLock method, but we didn't call it that since it's an overridable API.

Seems like _some_ change is needed here, but I can imagine a few very different options:

1. Leave this method as is; and add some heavy internal docing that it's effectively a NoLock method.
2. Change the doc comment to remove mention of it being called because of anything else, and just say the only difference between this method and the prior is one calls RaiseWorkspaceChanged, one doesn't.
3. Maybe switch the _serializationLock over to a monitor lock so we can just re-enter on the same thread, and the prior method can call this one without fear of deadlock. I'm not sure why _serializationLock is a SemaphoreSlim since it looks like all our acquisitions are synchronous.